keps/sig-node/5304-dra-attributes-downward-api/README.md

```diff
@@ -0,0 +1,830 @@
+
+## Motivation
+
+Workloads that need to interact with DRA-allocated devices (like KubeVirt virtual machines) require access to device-specific identifiers such as PCIe bus addresses or mediated device UUIDs. In order to fetch the attributes from allocated device, users first have to go to ResourceClaimStatus, find the request and device name, and then look up the resource slice with device name to get the attribute value. Ecosystem project like KubeVirt must resort to custom controllers that watch these objects and inject attributes via annotations/labels or other custom mechanisms, often leading to fragile, error-prone and racy designs.
+
+The Kubernetes Downward API provides a standard mechanism for exposing Pod and container metadata to workloads. Extending this API to support DRA device attributes would enable workloads to discover device information without requiring additional custom controllers or privileged access to the Kubernetes API.
+
+### Goals
+
+- Provide a stable Downward API path for device attributes associated with `pod.spec.resourceClaims[*]` requests
+- Support device attributes from `ResourceSlice` that are requested by user in pod spec
+- Maintain compatibility with existing DRA drivers without requiring changes to driver interfaces
+
+### Non-Goals
+
+- Expose the entirety of `ResourceClaim`/`ResourceSlice` objects in the Downward API
+- Allow arbitrary JSONPath into external objects via Downward API
+- Change or extend DRA driver interfaces
+- Support dynamic updates to device attributes after Pod container startup (for env vars)
+ - Propagate or snapshot attributes from `ResourceSlice` into `ResourceClaim` at allocation time (no scheduler involvement)
+
+## Proposal
+
+This proposal introduces a new Downward API selector (`resourceSliceAttributeRef`) that allows Pods to reference DRA device attributes in environment variables. The kubelet will implement a local DRA attributes controller that:
+
+1. Watches Pods scheduled to the node and identifies those with `pod.spec.resourceClaims`
+2. Watches `ResourceClaim` objects in the Pod's namespace to retrieve allocation information
+3. Watches `ResourceSlice` objects for the node and driver to resolve device attributes
+4. Maintains a per-Pod cache of `(claimName, requestName) -> {attribute: value}` mappings
```

Comment:
This has a serious scalability impact

Reply:
I don't think we want the kubelet to be in the business of watching all `ResourceSlices`. We would need a way of allowing the kubelet to quickly look up the device in a specific resource slice on-demand (possibly caching it upon first look up). Either that, or make sure that it only opened a watch for `ResourceSlices` matching the current node.